src/job/jobHandler.ts

```diff
@@ -234,21 +234,90 @@ export abstract class JobHandler {
+    const stages = {
+      ['get-build-dependencies']: 'buildDepsExe',
+      ['next-gen-parse']: 'parseExe',
+      ['next-gen-html']: 'htmlExe',
+      ['oas-page-build']: 'oasPageBuildExe',
+    };
+
+    // get the prerequisite commands which should be all commands up to `rm -f makefile`
+    const endOfPrerequisiteCommands = this.currJob.buildCommands.indexOf('rm -f makefile');
+    const index = endOfPrerequisiteCommands + 1;
+    const prerequisiteCommands = this.currJob.buildCommands.slice(0, index);
+    await this._logger.save(this.currJob._id, `'(PREREQUISITE COMMANDS)'${prerequisiteCommands.join(' && ')}`);
+
+    const makeCommands = this.currJob.buildCommands.slice(index);
+    await this._logger.save(this.currJob._id, `'(MAKE COMMANDS)'${makeCommands.join(' && ')}`);
+
+    // call prerequisite commands
+    const prerequisiteResp = await this._commandExecutor.execute(prerequisiteCommands);
+    await this.loggingMessage(prerequisiteResp);
+
+    for (const command of makeCommands) {
+      // works for any make command with the following signature make <make-rule>
+      const key = command.split(' ')[1].trim();
+      if (stages[key]) {
+        const makeCommandsWithBenchmarksResponse = await this.callWithBenchmark(command, stages[key]);
+        await this.loggingMessage(makeCommandsWithBenchmarksResponse);
+      } else {
+        const makeCommandsResp = await this._commandExecutor.execute([command]);
+        await this.loggingMessage(makeCommandsResp);
+      }
+    }
+    await this._logger.save(this.currJob._id, `${'(BUILD)'.padEnd(15)}Finished Build`);
+  }
+
+  private async exeBuild(): Promise<void> {
+    const resp = await this._commandExecutor.execute(this.currJob.buildCommands);
+    this.loggingMessage(resp);
+    await this._logger.save(this.currJob._id, `${'(BUILD)'.padEnd(15)}Finished Build`);
+  }
+
   @throwIfJobInterupted()
   private async executeBuild(): Promise<boolean> {
     if (this.currJob.buildCommands && this.currJob.buildCommands.length > 0) {
       await this._logger.save(this.currJob._id, `${'(BUILD)'.padEnd(15)}Running Build`);
       await this._logger.save(this.currJob._id, `${'(BUILD)'.padEnd(15)}running worker.sh`);
-      const resp = await this._commandExecutor.execute(this.currJob.buildCommands);
-      await this._logger.save(this.currJob._id, `${'(BUILD)'.padEnd(15)}Finished Build`);
-      await this._logger.save(
-        this.currJob._id,
-        `${'(BUILD)'.padEnd(15)}worker.sh run details:\n\n${resp.output}\n---\n${resp.error}`
-      );
-      if (resp.status != 'success') {
-        const error = new AutoBuilderError(resp.error, 'BuildError');
-        await this.logError(error);
-        throw error;
+      // Using indexOf to handle between the two environments.
+      const insertionPoint = this.currJob.buildCommands.indexOf('make next-gen-html');
+      if (insertionPoint !== -1) {
+        this.currJob.buildCommands.splice(insertionPoint, 0, 'make next-gen-parse');
```

Comment:
@rayangler my thought was to keep the changes within one file or method. This way if the `prepStageSpecificNextGenCommands` order needed to change the `executeBuild` would be flexible enough to adapt if need be.

Reply:
That makes sense. I'm not too sure how flexible keeping this logic in `executeBuild` would be though. If in the future we want to split out the call to the persistence module as a separate command, we would definitely need to add a new build command to accommodate that, and this logic would still need to change. My preference in terms of readability would be to have the commands listed out together, rather than having them be inserted elsewhere if they don't need to be.

If it makes more sense to separate it out like this, a short comment explaining why the commands are split like this would probably be appreciated by future us!

Reply:
@rayangler good callout. I will keep the commands together and move them into the `prepStageSpecificNextGenCommands` methods.